File: layers/src/main/java/com/psliusar/layers/RemoveTransition.java
package com.psliusar.layers;

import android.os.Bundle;

import java.util.ArrayList;

import androidx.annotation.NonNull;

public class RemoveTransition<LAYER extends Layer<?>> extends Transition<LAYER> {

    private int lowestVisibleLayer = -1;

    RemoveTransition(@NonNull Layers layers, int index) {
        super(layers, index);
    }

    @NonNull
    @Override
    public Transition<LAYER> setArguments(@NonNull Bundle arguments) {
        throw new IllegalArgumentException("Unable to set arguments when removing layer");
    }

    @NonNull
    @Override
    public Transition<LAYER> setName(@NonNull String name) {
        throw new IllegalArgumentException("Unable to set name when removing layer");
    }

    @NonNull
    @Override
    public Transition<LAYER> setInAnimation(int outAnimId, int inAnimId) {
        throw new IllegalArgumentException("Unable to set intro animation when removing layer");
    }

    @Override
    protected void onTransition() {
        super.onTransition();

        // TODO remove layer with custom index

        stackEntry.inTransition = true;

        final int stackSize = layers.getStackSize();
        lowestVisibleLayer = layers.getLowestVisibleLayer();

        if (index >= lowestVisibleLayer) {
            // Mark layers for transition
            setTransitionState(lowestVisibleLayer);

            // Make sure we have all required views in layout
            layers.ensureViews();

            // Animate layers
            for (int i = lowestVisibleLayer; i < stackSize; i++) {
                animateLayer(layers.getStackEntryAt(i).layerInstance,
                    i == index ? AnimationType.ANIMATION_UPPER_OUT : AnimationType.ANIMATION_LOWER_IN);
            }
        }

        stackEntry.valid = false;
    }

    @Override
    protected void onAfterTransition() {
        super.onAfterTransition();

        if (index >= lowestVisibleLayer) {
            resetTransitionState(lowestVisibleLayer);
        }
        layers.removeLayerAt(index);
    }

    @Override
    protected void fastForward(@NonNull ArrayList<StackEntry> stack) {
        if (!isApplied()) {
            stack.remove(index); // size should be checked earlier
        }
    }
}

File: layers/src/main/java/com/psliusar/layers/RemoveTransition.kt
package com.psliusar.layers

import android.os.Bundle
import java.util.ArrayList

/**
 * Transition removes [Layer] from the stack at the given index. Removes from the top of the stack
 * if the index is omitted.
 */
internal class RemoveTransition<L : Layer>(
    layers: Layers,
    index: Int
) : Transition<L>(layers, layers.getStackEntryAt(index)) {

    private var lowestVisibleLayer = -1

    init {
        this.index = index
    }

    override var arguments: Bundle?
        get() = super.arguments
        set(_) {
            throw IllegalArgumentException("Unable to set arguments when removing layer")
        }

    override var name: String?
        get() = super.name
        set(_) {
            throw IllegalArgumentException("Unable to set name when removing layer")
        }

    override fun setInAnimation(outAnimId: Int, inAnimId: Int): Transition<L> {
        throw IllegalArgumentException("Unable to set intro animation when removing layer")
    }

    override fun onTransition() {
        stackEntry.inTransition = true

        val stackSize = layers.stackSize
        lowestVisibleLayer = layers.getLowestVisibleLayer()
        stackEntry.valid = false

        if (index < lowestVisibleLayer) return

        // Mark layers for transition
        setTransitionState(lowestVisibleLayer)

        // Make sure we have all required views in layout for animation
        layers.ensureViews()

        // Animate layers
        for (i in lowestVisibleLayer until stackSize) {
            val layer = layers.getStackEntryAt(i).layerInstance
                ?: throw IllegalStateException("Layer instance must exist")
            val anim = when {
                i == index -> AnimationType.ANIMATION_UPPER_OUT
                i < index -> AnimationType.ANIMATION_LOWER_IN
                else -> continue
            }
            animateLayer(layer, anim)
        }
    }

    override fun onAfterTransition() {
        if (index >= lowestVisibleLayer) {
            resetTransitionState(lowestVisibleLayer)
        }
        layers.removeLayerAt<Layer>(index)
    }

    override fun fastForward(stack: ArrayList<StackEntry>) {
        if (started) return

        stack.removeAt(index) // size should be checked earlier
    }
}